platform/asterix/boot/src/drivers/flash/mx25u.c

```diff
@@ -1,81 +1,114 @@
+  config.prot_if.writeoc = NRF_QSPI_WRITEOC_PP4O;
+  config.phy_if.sck_freq = NRF_QSPI_FREQ_DIV4;
+  config.prot_if.addrmode = NRF_QSPI_ADDRMODE_32BIT;
+
+  nrfx_qspi_init(&config, NULL, NULL);
 
-void flash_init(void)
-{
-	nrf_qspi_pins_t pins;
-	nrf_qspi_pins_t disconnected_pins = {
-		.sck_pin = NRF_QSPI_PIN_NOT_CONNECTED,
-		.csn_pin = NRF_QSPI_PIN_NOT_CONNECTED,
-		.io0_pin = NRF_QSPI_PIN_NOT_CONNECTED,
-		.io1_pin = NRF_QSPI_PIN_NOT_CONNECTED,
-		.io2_pin = NRF_QSPI_PIN_NOT_CONNECTED,
-		.io3_pin = NRF_QSPI_PIN_NOT_CONNECTED,
-	};
-
-	config.prot_if.readoc = NRF_QSPI_READOC_READ4IO;
-	config.prot_if.writeoc = NRF_QSPI_WRITEOC_PP4IO;
-	config.phy_if.sck_freq = NRF_QSPI_FREQ_DIV4;
-	config.phy_if.sck_delay = 0x00;
-	// config.skip_gpio_cfg = true;
-	config.skip_psel_cfg = true;
-
-	nrfx_qspi_init(&config, NULL, NULL);
-
-	nrf_qspi_pins_set(NRF_QSPI, &disconnected_pins);
-	nrfx_qspi_activate(true);
-	nrf_qspi_pins_set(NRF_QSPI, &config.pins);
-
-	/* exit DPD */
-	nrf_qspi_cinstr_conf_t instr_config = {
-		.opcode = SPI_NOR_CMD_RDPD,
-		.length = NRF_QSPI_CINSTR_LEN_1B,
-		.io2_level = true,
-		.io3_level = true,
-		.wipwait = false,
-		.wren = false,
-	};
-
-	nrfx_qspi_cinstr_xfer(&instr_config, NULL, NULL);
-
-	NRFX_DELAY_US(35);
+  // reset the flash
+  prv_write_register(SPI_NOR_CMD_ENRST, NULL, 0);
+  prv_write_register(SPI_NOR_CMD_RST, NULL, 0);
+  NRFX_DELAY_US(35);
+
+  // enable 4 byte addressing
+  prv_write_register(SPI_NOR_CMD_EN4B, NULL, 0);
+
+  // enable QE bit
+  prv_read_register(SPI_NOR_CMD_RDSR1, &sr[0], 1);
+  prv_read_register(SPI_NOR_CMD_RDSR2, &sr[1], 1);
+  sr[1] |= (1 << 1);
+  prv_write_register(SPI_NOR_CMD_WRSR, sr, 2);
 }
 
-bool flash_sanity_check(void)
-{
-	uint8_t buf[3];
-	uint32_t id;
-	nrf_qspi_cinstr_conf_t instr_config = {
-		.opcode = SPI_NOR_CMD_RDID,
-		.length = NRF_QSPI_CINSTR_LEN_4B,
-		.io2_level = true,
-		.io3_level = true,
-		.wipwait = false,
-		.wren = false,
-	};
-
-	nrfx_qspi_cinstr_xfer(&instr_config, NULL, buf);
-
-	id = buf[0] << 16 | buf[1] << 8 | buf[2];
-	dbgserial_print("JEDEC ID: ");
-	dbgserial_print_hex(id);
-	dbgserial_newline();
-
-	return (id == BOARD_FLASH_JEDEC_ID);
+bool flash_sanity_check(void) {
+  uint8_t buf[3];
+  uint32_t id;
+
+  prv_read_register(SPI_NOR_CMD_RDID, buf, sizeof(buf));
+
+  id = buf[0] << 16 | buf[1] << 8 | buf[2];
+  dbgserial_print("JEDEC ID: ");
+  dbgserial_print_hex(id);
+  dbgserial_newline();
+
+  return (id == BOARD_FLASH_JEDEC_ID);
 }
 
-void flash_read_bytes(uint8_t *buffer_ptr, uint32_t start_addr, uint32_t buffer_size)
-{
-	nrfx_qspi_read(buffer_ptr, buffer_size, start_addr);
+void flash_read_bytes(uint8_t *buffer_ptr, uint32_t start_addr, uint32_t buffer_size) {
+  uint8_t __attribute__((aligned(4))) b_buf[4];
+  uint8_t buf_pre;
+  uint8_t buf_suf;
+  uint32_t buf_mid;
+  nrfx_err_t err;
+
+  buf_pre = (4U - (uint8_t)((uint32_t)buffer_ptr % 4U)) % 4U;
+  if (buf_pre > buffer_size) {
+    buf_pre = buffer_size;
+  }
```

Comment:
another option would be to disallow this entirely and assert :) but this is fine too

Reply:
The problem is that firmware images are by default not size-aligned, so we'd need to enforce that at build stage by adding extra padding. Since the bootloader is the immutable piece of the system, I'd play safe and support any alignment for now.